File: .skills/skills/security-auditor/scripts/scan-secrets.js
#!/usr/bin/env node
/**
 * Scan repository for leaked secrets, API keys, and sensitive data.
 * Part of the security-auditor skill (D-016).
 *
 * Usage: node scan-secrets.js [--format json|text]
 */

const fs = require("fs");
const path = require("path");
const { execSync } = require("child_process");

const REPO_ROOT = process.env.REPO_ROOT || process.cwd();

const SECRET_PATTERNS = [
  [/sb_secret_[A-Za-z0-9_-]+/g, "Supabase service_role key"],
  [/service_role["'\s:=]+[A-Za-z0-9_.+-]{20,}/g, "service_role key assignment"],
  [/(?:password|passwd|pwd)\s*[=:]\s*["'][^"']{4,}/gi, "Hardcoded password"],
  [/(?:api[_-]?key)\s*[=:]\s*["'][^"']{10,}/gi, "API key"],
  [/Bearer\s+[A-Za-z0-9_.-]{20,}/g, "Bearer token"],
  [/eyJhbGciOi[A-Za-z0-9_-]{50,}/g, "JWT token"],
];

const SKIP_DIRS = new Set([
  "node_modules", ".next", ".git", "test-results", "playwright-report",
  "__pycache__", ".cache", "dist", "build", ".vercel",
]);
const SKIP_EXTENSIONS = new Set([
  ".png", ".jpg", ".jpeg", ".gif", ".ico", ".svg", ".woff", ".woff2",
  ".ttf", ".eot", ".mp4", ".mp3", ".zip", ".tar", ".gz", ".lock",
]);
const ALLOWED_FILES = new Set([".env.local", ".env.development.local", ".env.production.local"]);

function getTrackedFiles() {
  try {
    const out = execSync("git ls-files", { cwd: REPO_ROOT, encoding: "utf-8" });
    return new Set(out.trim().split("\n").filter(Boolean));
  } catch { return new Set(); }
}

function scanFile(filePath, trackedFiles) {
  const findings = [];
  const relPath = path.relative(REPO_ROOT, filePath).replace(/\\/g, "/");

  if (!trackedFiles.has(relPath)) return findings;
  const ext = path.extname(filePath).toLowerCase();
  if (SKIP_EXTENSIONS.has(ext)) return findings;
  if (ALLOWED_FILES.has(path.basename(filePath))) return findings;

  let content;
  try { content = fs.readFileSync(filePath, "utf-8"); }
  catch { return findings; }

  for (const [pattern, description] of SECRET_PATTERNS) {
    // Reset regex lastIndex
    pattern.lastIndex = 0;
    let match;
    while ((match = pattern.exec(content)) !== null) {
      const line = content.substring(0, match.index).split("\n").length;
      const matched = match[0];
      const display = matched.length > 20
        ? matched.substring(0, 10) + "..." + matched.substring(matched.length - 5)
        : matched;
      findings.push({ file: relPath, line, type: description, match: display, severity: "CRITICAL" });
    }
  }
  return findings;
}

function walkDir(dir, trackedFiles) {
  let findings = [];
  let filesScanned = 0;
  const entries = fs.readdirSync(dir, { withFileTypes: true });

  for (const entry of entries) {
    if (SKIP_DIRS.has(entry.name)) continue;
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      const sub = walkDir(fullPath, trackedFiles);
      findings = findings.concat(sub.findings);
      filesScanned += sub.filesScanned;
    } else {
      const relPath = path.relative(REPO_ROOT, fullPath).replace(/\\/g, "/");
      if (trackedFiles.has(relPath)) filesScanned++;
      findings = findings.concat(scanFile(fullPath, trackedFiles));
    }
  }
  return { findings, filesScanned };
}

function checkGitignore() {
  const gitignorePath = path.join(REPO_ROOT, ".gitignore");
  if (!fs.existsSync(gitignorePath))
    return { status: "FAIL", reason: ".gitignore not found" };
  const content = fs.readFileSync(gitignorePath, "utf-8");
  if (content.includes(".env.local") || content.includes(".env*.local"))
    return { status: "PASS" };
  return { status: "FAIL", reason: ".env.local not in .gitignore" };
}

function checkEnvInHistory() {
  try {
    const out = execSync(
      "git log --all --diff-filter=A --name-only --pretty=format: -- \"*.env*\"",
      { cwd: REPO_ROOT, encoding: "utf-8" }
    );
    return out.trim().split("\n")
      .filter(f => f.trim() && !f.endsWith(".example") && !f.endsWith(".localecho"))
      .map(f => f.trim());
  } catch { return []; }
}

// --- Main ---
const format = process.argv.includes("--format")
  ? process.argv[process.argv.indexOf("--format") + 1]
  : "text";

const trackedFiles = getTrackedFiles();
const { findings, filesScanned } = walkDir(REPO_ROOT, trackedFiles);
const gitignore = checkGitignore();
const envInHistory = checkEnvInHistory();
const status = findings.length > 0 || gitignore.status === "FAIL" ? "CRITICAL" : "PASS";

const result = { files_scanned: filesScanned, secrets_found: findings.length, findings, gitignore, env_in_history: envInHistory, status };

if (format === "json") {
  console.log(JSON.stringify(result, null, 2));
} else {
  console.log(`Files scanned: ${filesScanned}`);
  console.log(`Secrets found: ${findings.length}`);
  console.log(`.gitignore: ${gitignore.status}`);
  console.log(`Env files in git history: ${envInHistory.length}`);
  if (findings.length) {
    console.log("\n--- FINDINGS ---");
    for (const f of findings) console.log(`  CRITICAL: ${f.file}:${f.line} — ${f.type} (${f.match})`);
  }
  if (envInHistory.length) {
    console.log("\n--- ENV FILES IN HISTORY ---");
    for (const f of envInHistory) console.log(`  WARNING: ${f}`);
  }
  console.log(`\nVerdict: ${status}`);
}

process.exit(status === "PASS" ? 0 : 1);

File: .skills/skills/security-auditor/scripts/scan-secrets.py
#!/usr/bin/env python3
"""
Scan repository for leaked secrets, API keys, and sensitive data.
Part of the security-auditor skill (D-016).

Usage:
  python scan-secrets.py <repo_root> [--format json|text]
"""

import argparse
import json
import os
import re
import subprocess
import sys

# Patterns that indicate secrets in source code
SECRET_PATTERNS = [
    (r'sb_secret_[A-Za-z0-9_-]+', 'Supabase service_role key'),
    (r'service_role["\s:=]+[A-Za-z0-9_.-]{20,}', 'service_role key assignment'),
    (r'(?i)password\s*[=:]\s*["\'][^"\']{4,}', 'Hardcoded password'),
    (r'(?i)api[_-]?key\s*[=:]\s*["\'][^"\']{10,}', 'API key'),
    (r'Bearer\s+[A-Za-z0-9_.-]{20,}', 'Bearer token'),
    (r'(?i)secret[_-]?key\s*[=:]\s*["\'][^"\']{10,}', 'Secret key'),
    (r'eyJhbGciOi[A-Za-z0-9_-]{50,}', 'JWT token'),
]

# Files/dirs to skip
SKIP_DIRS = {
    'node_modules', '.next', '.git', 'test-results', 'playwright-report',
    '__pycache__', '.cache', 'dist', 'build', '.vercel'
}
SKIP_EXTENSIONS = {
    '.png', '.jpg', '.jpeg', '.gif', '.ico', '.svg', '.woff', '.woff2',
    '.ttf', '.eot', '.mp4', '.mp3', '.zip', '.tar', '.gz', '.lock'
}
# Files that are EXPECTED to contain keys (not tracked by git)
ALLOWED_FILES = {'.env.local', '.env.development.local', '.env.production.local'}


def get_tracked_files(repo_root):
    """Get list of files tracked by git."""
    try:
        result = subprocess.run(
            ['git', 'ls-files'],
            cwd=repo_root, capture_output=True, text=True, check=True
        )
        return set(result.stdout.strip().split('\n'))
    except subprocess.CalledProcessError:
        return set()


def scan_file(filepath, tracked_files, repo_root):
    """Scan a single file for secret patterns."""
    findings = []
    rel_path = os.path.relpath(filepath, repo_root).replace('\\', '/')

    # Skip if not tracked by git (secrets in .env.local are expected)
    if rel_path not in tracked_files:
        return findings

    # Skip binary/media files
    _, ext = os.path.splitext(filepath)
    if ext.lower() in SKIP_EXTENSIONS:
        return findings

    # Skip allowed files
    basename = os.path.basename(filepath)
    if basename in ALLOWED_FILES:
        return findings

    try:
        with open(filepath, 'r', encoding='utf-8', errors='ignore') as f:
            content = f.read()
    except (OSError, UnicodeDecodeError):
        return findings

    for pattern, description in SECRET_PATTERNS:
        for match in re.finditer(pattern, content):
            # Get line number
            line_num = content[:match.start()].count('\n') + 1
            # Truncate the match for display (don't show full secrets)
            matched_text = match.group()
            if len(matched_text) > 20:
                matched_text = matched_text[:10] + '...' + matched_text[-5:]
            findings.append({
                'file': rel_path,
                'line': line_num,
                'type': description,
                'match': matched_text,
                'severity': 'CRITICAL'
            })

    return findings


def check_gitignore(repo_root):
    """Verify .env.local is in .gitignore."""
    gitignore_path = os.path.join(repo_root, '.gitignore')
    if not os.path.exists(gitignore_path):
        return {'status': 'FAIL', 'reason': '.gitignore not found'}

    with open(gitignore_path, 'r') as f:
        content = f.read()

    patterns_to_check = ['.env.local', '.env*.local']
    for pattern in patterns_to_check:
        if pattern in content:
            return {'status': 'PASS', 'pattern': pattern}

    return {'status': 'FAIL', 'reason': '.env.local not found in .gitignore'}


def check_env_in_git_history(repo_root):
    """Check if .env files were ever committed."""
    try:
        result = subprocess.run(
            ['git', 'log', '--all', '--diff-filter=A', '--name-only', '--pretty=format:', '--', '*.env*'],
            cwd=repo_root, capture_output=True, text=True, check=True
        )
        files = [f.strip() for f in result.stdout.strip().split('\n') if f.strip()]
        # Filter out .env.example which is expected
        suspicious = [f for f in files if f and not f.endswith('.example') and not f.endswith('.localecho')]
        return suspicious
    except subprocess.CalledProcessError:
        return []


def main():
    parser = argparse.ArgumentParser(description='Scan for secrets in repository')
    parser.add_argument('repo_root', help='Path to repository root')
    parser.add_argument('--format', choices=['json', 'text'], default='text')
    args = parser.parse_args()

    repo_root = os.path.abspath(args.repo_root)
    tracked_files = get_tracked_files(repo_root)
    all_findings = []
    files_scanned = 0

    # Scan all tracked files
    for root, dirs, files in os.walk(repo_root):
        # Skip excluded directories
        dirs[:] = [d for d in dirs if d not in SKIP_DIRS]

        for filename in files:
            filepath = os.path.join(root, filename)
            findings = scan_file(filepath, tracked_files, repo_root)
            if findings:
                all_findings.extend(findings)
            rel = os.path.relpath(filepath, repo_root).replace('\\', '/')
            if rel in tracked_files:
                files_scanned += 1

    # Check .gitignore
    gitignore_check = check_gitignore(repo_root)

    # Check git history
    env_in_history = check_env_in_git_history(repo_root)

    result = {
        'files_scanned': files_scanned,
        'secrets_found': len(all_findings),
        'findings': all_findings,
        'gitignore': gitignore_check,
        'env_in_history': env_in_history,
        'status': 'CRITICAL' if all_findings or gitignore_check['status'] == 'FAIL' else 'PASS'
    }

    if args.format == 'json':
        print(json.dumps(result, indent=2))
    else:
        print(f"Files scanned: {files_scanned}")
        print(f"Secrets found: {len(all_findings)}")
        print(f".gitignore: {gitignore_check['status']}")
        print(f"Env files in git history: {len(env_in_history)}")
        if all_findings:
            print("\n--- FINDINGS ---")
            for f in all_findings:
                print(f"  CRITICAL: {f['file']}:{f['line']} — {f['type']} ({f['match']})")
        if env_in_history:
            print(f"\n--- ENV FILES IN HISTORY ---")
            for f in env_in_history:
                print(f"  WARNING: {f}")
        print(f"\nVerdict: {result['status']}")

    sys.exit(0 if result['status'] == 'PASS' else 1)


if __name__ == '__main__':
    main()
